jdbc/src/main/java/org/springframework/jdbc/core/JdbcTemplate.java

```diff
@@ -24,50 +21,29 @@ public JdbcTemplate(final DataSource dataSource) {
     }
 
     public int update(String sql, Object... args) {
+        log.debug("query : {}", sql);
+        return execute(new SetPreparedStatementMaker(sql, args), new PreparedStatementUpdateExecuter());
+    }
+
+    private <T> T execute(
+            PreparedStatementMaker pstmtMaker,
+            PreparedStatementExecuter<T> pstmtExecuter
+    ) {
         try (
                 Connection conn = dataSource.getConnection();
-                PreparedStatement pstmt = conn.prepareStatement(sql)
+                PreparedStatement pstmt = pstmtMaker.makePreparedStatement(conn)
         ) {
-            log.debug("query : {}", sql);
-
-            setParameters(pstmt, args);
-
-            return pstmt.executeUpdate();
+            return pstmtExecuter.execute(pstmt);
         } catch (SQLException e) {
-            log.error(e.getMessage(), e);
-            throw new RuntimeException(e);
-        }
-    }
-
-    private void setParameters(PreparedStatement pstmt, Object[] args) throws SQLException {
-        for (int i = 0; i < args.length; i++) {
-            pstmt.setObject(i + 1, args[i]);
+            throw new DataAccessException(e);
         }
     }
 
     public <T> List<T> query(String sql, RowMapper<T> rowMapper, Object... args) {
-        try (
-                Connection conn = dataSource.getConnection();
-                PreparedStatement pstmt = conn.prepareStatement(sql)
-        ) {
-            log.debug("query : {}", sql);
-
-            setParameters(pstmt, args);
-
-            try (ResultSet resultSet = pstmt.executeQuery()) {
-                List<T> results = new ArrayList<>();
-                while (resultSet.next()) {
-                    results.add(rowMapper.mapRow(resultSet));
-                }
-                return results;
-            }
-        } catch (SQLException e) {
-            log.error(e.getMessage(), e);
-            throw new RuntimeException(e);
-        }
+        log.debug("query : {}", sql);
+        return execute(new SetPreparedStatementMaker(sql, args), new PreparedStatementQueryExecuter<>(rowMapper));
     }
 
-    @Nullable
     public <T> T queryForObject(String sql, RowMapper<T> rowMapper, Object... args) {
         final List<T> objects = query(sql, rowMapper, args);
         return getSingleObject(objects);
```

Comment:
execute로 통일함에 있어 queryForObject도 함께 묶이지 않을까 싶었는데 queryForObject만 별도의 흐름을 가지고있군요..!
혹시 이렇게 구성한 이유가 있을까요?

추상화 레벨을 더 높여서 queryForObject도 execute() 메소드로 묶을 수는 없을까요?

```suggestion
    public <T> T queryForObject(String sql, RowMapper<T> rowMapper, Object... args) {
        return execute(new SetPreparedStatementMaker(sql, args), new PreparedStatementQueryForObjectExecuter<>(rowMapper));
```

Reply:
주노가 제안해주신대로 PreparedStatementQueryForObjectExecuter 객체를 생성해도 괜찮을 것 같아요.

다만 저는 아래와 같은 이유로 현재의 구조를 유지하고 싶습니다.

- PreparedStatementExecuter의 구현체가 예외를 관리하는 문제 : 추상화된 인터페이스 이름에서 알 수 있듯이, 이 인터페이스의 구현체들은 단순히 PreparedStatement 의 execute() 관련 메서드를 실행하는 책임을 갖습니다. 만약 execute() 이후에 예외까지 처리한다면, 부여된 책임을 넘어서는 것이라 생각합니다.

- 중복 코드 : 만약 PreparedStatementQueryForObjectExecuter 클래스를 작성한다면, 그 구현 코드는 PreparedStatementQueryExecuter 와 비슷하거나 중복될 것입니다. queryForObject 메서드 내부에서 query 메서드를 호출하는 것은 중복 코드를 줄이는 방법 중 하나로 택한 것입니다. 따라서 queryForObject를 위한 클래스를 작성한다면 중복 코드가 생길 수밖에 없다고 생각합니다.

주노는 어떻게 생각하시나요?

Reply:
후추의 의견도 일리있군요!!

다만 저는 개인적으로 public method인 `query()`를 재사용하는 것이 적절한지 의문이 듭니다..
만약 query 메소드에 대해 어떠한 추가 요구사항이 반영되어야한다면 연관된 `queryForObject()`도 함께 영향을 받게 된다고 생각합니다.

추후 관리 측면에서 public method끼리의 강한 결합이 어떤 문제를 일으키지는 않을까? 라는 의구심이 자꾸 맴도네요 😅

또한 아래와 같이 구현체가 예외를 상위레벨로 넘겨준다는 것은 어떻게 생각하시나요?
이렇게 하면 외부에서 SQLException을 catch하고 있으므로 충분히 핸들링 할 여지를 외부로 넘긴다고 생각합니다.

```java
// 만약.. 구현한다면..?
public class PreparedStatementQueryForObjectExecuter<T> implements PreparedStatementExecuter<T> {

    private final RowMapper<T> rowMapper;

    public PreparedStatementQueryForObjectExecuter(RowMapper<T> rowMapper) {
        this.rowMapper = rowMapper;
    }

    @Override
    public T execute(PreparedStatement pstmt) throws SQLException {
        try (ResultSet resultSet = pstmt.executeQuery()) {
            T result = null;

            while (resultSet.next()) {
                if (result != null) {
                    throw new SQLException("조회 데이터가 한 개 이상 존재합니다.");
                }
                result = rowMapper.mapRow(resultSet);
            }

            if (result == null) {
                throw new SQLException("조회 데이터가 존재하지 않습니다.");
            }

            return result;
        }
    }
}
```

Reply:
친절하고도 예리한 리뷰 정말 감사드립니다!

덕분에 재미있는 고민을 할 수 있었어요.

## public 메서드들끼리 강결합

저는 한 클래스에서 public 메서드 결합이 있어도 괜찮다고 생각합니다. 아래 Reader 클래스 예시를 통해 제 생각을 말씀드려 볼게요.

```java
public class Reader {

    public void read(String message) {
        System.out.println(message);
    }

    public void readTwice(String message) {
        System.out.println(message);
        System.out.println(message);
    }
}
```
Reader 클래스는 message를 받아 console에 출력해주는 메서드를 갖고 있어요. 다만, System.out.println() 메서드가 중복되고 있습니다. 이를 아래와 같이 수정할 수 있어요.

```java
public class Reader {

    public void read(String message) {
        readOnce(message);
    }

    private void readOnce(String message) {
        System.out.println(message);
    }

    public void readTwice(String message) {
        readOnce(message);
        readOnce(message);
    }
}
```
중복되는 메서드를 private 메서드로 추출해서 사용하고 있어요. 그런데 이는 아래와 다를 바가 없습니다!

```java
public class Reader {

    public void read(String message) {
        System.out.println(message);
    }

    public void readTwice(String message) {
        read(message);
        read(message);
    }
}
``` 

이러한 점에서 볼 때, 만약 한 public 메서드에서 다른 public 메서드를 호출하는 것이 문제라면, 여러 public 메서드에서 동일한 private 메서드를 사용하는 것도 문제입니다. 

그러나 사실 private 메서드를 여러 곳에서 사용하는 것은 문제가 아니라, 지향해야 할 바이죠!

따라서 public 메서드 사이에 강결합도 문제가 아니라고 생각합니다.

## PreparedStatementQueryForObjectExecuter 구현하면?
주노가 작성해주신 코드 잘 보았습니다. 그 코드를 제 입맛에 맞게 구현하면 다음과 같았어요. (result를 null로 초기화 해두었다가 값이 달라지는 경우에 따라 로직을 수행하기보다는, List에 값을 담아두는 것이 낫다고 판단했어요.)

```java
public class PreparedStatementQueryForObjectExecuter<T> implements PreparedStatementExecuter<T> {

    private static final int SINGLE_SIZE = 1;

    private final RowMapper<T> rowMapper;

    public PreparedStatementQueryForObjectExecuter(RowMapper<T> rowMapper) {
        this.rowMapper = rowMapper;
    }

    @Override
    public T execute(PreparedStatement pstmt) throws SQLException {
        try (ResultSet resultSet = pstmt.executeQuery()) {
            List<T> results = new ArrayList<>();
            while (resultSet.next()) {
                results.add(rowMapper.mapRow(resultSet));
            }
            return getSingleObject(results);
        }
    }

    private T getSingleObject(List<T> objects) throws SQLException {
        validateEmpty(objects);
        validateSingleSize(objects);
        return objects.iterator().next();
    }

    private void validateEmpty(List<T> objects) throws SQLException {
        if (objects.isEmpty()) {
            throw new SQLException("조회 데이터가 존재하지 않습니다.");
        }
    }

    private void validateSingleSize(List<T> objects) throws SQLException {
        if (objects.size() > SINGLE_SIZE) {
            throw new SQLException("조회 데이터가 한 개 이상 존재합니다.");
        }
    }
}
```

그러나 execute 메서드의 내부 로직은 PreparedStatementQueryExecuter 클래스의 구현부와 너무나도 흡사합니다.

![image](https://github.com/woowacourse/jwp-dashboard-jdbc/assets/106813090/bcd5f42e-7f9e-4204-b57b-cf24c4fe13f9)

결국 주노가 짜주신 로직도 중복되지 않은 것처럼 보이는, 중복이라고 생각해요. 그래서 이를 줄이는 선택을 하고 싶습니다.

## 결론

- public method 끼리의 결합이 문제가 되지 않을 것으로 판단합니다.
- 코드의 중복을 줄이고 싶습니다.
- 따라서 기존의 방식을 유지하겠습니다.

주노는 어떻게 생각하시는지 궁금합니다.

Reply:
오! 후추의 의견을 쭉 읽다보니 애매하게 가려웠던 부분이 해소된것 같아요 😊
public 메소드의 강결합에 대한 명확한 예시와 의견을 제시해주시니 이해가 한결 수월했던 것 같습니다 👍
또한 코드의 중복을 줄이기 위해 `PreparedStatementQueryForObjectExecuter` 대신 비슷한 구조의 `PreparedStatementQueryExecuter`를 사용하는것도 납득이 되었습니다!!

후추의 생각에 대해 많이 배울 수 있는 즐거운 토론이였습니다 😆